files/acl.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+namespace Automattic\VIP\Files\Acl;
+
+const FILE_IS_PUBLIC = 'FILE_IS_PUBLIC';
+const FILE_IS_PRIVATE_AND_ALLOWED = 'FILE_IS_PRIVATE_AND_ALLOWED';
+const FILE_IS_PRIVATE_AND_DENIED = 'FILE_IS_PRIVATE_AND_DENIED';
+const FILE_NOT_FOUND = 'FILE_NOT_FOUND';
+
+/**
+ * Validate the incoming files request.
+ * 
+ * Note: As the function name suggests, this is called before WordPress is loaded.
+ *
+ * @param string $file_request_uri
+ */
+function pre_wp_validate_request( $file_request_uri ) {
+	// Note: cannot use WordPress functions here.
+
+	if ( ! $file_request_uri ) {
+		return [ 400, 'missing-uri' ];
+	}
+
+	// strpos since we can have subsite / subdirectory paths.
+	if ( false === strpos( $file_request_uri, '/wp-content/uploads/' ) ) {
+		return [ 400, 'invalid-path' ];
+	}
+
+	// Strip off non-path elements (scheme/host/querystring).
+	$file_path = parse_url( $file_request_uri, PHP_URL_PATH );
+	if ( ! $file_path ) {
+		return [ 400, 'parse-error' ];
+	}
+
+	if ( 0 !== strpos( $file_path, '/' ) ) {
+		return [ 400, 'relative-path' ];
+	}
+
+	// Strip off subdirectory (i.e. /en/wp-content/ <= remove `/en` )
+	$wp_content_index = strpos( $file_path, '/wp-content/uploads/' );
+	$file_path = substr( $file_path, $wp_content_index + strlen( '/wp-content/uploads/' ) );
+
+	return $file_path;
+}
+
+/**
+ * Helper function to send error headers.
+ *
+ * Called before WordPress is loaded.
+ *
+ * @param array $error Array containing [ (int) status code, (string) error message ]
+ */
+function pre_wp_send_error_headers( $error ) {
+	list( $error_code, $error_message ) = $error;
+
+	http_response_code( $error_code );
+
+	header( 'X-Reason: ' . $error_message );
```

Comment:
The `X-` prefix is [deprecated](https://tools.ietf.org/html/rfc6648), and I think we could give a bit more context to the header name by calling it `Acl-Reason` or similar.

Reply:
Interesting; TIL. The intent of that seems more geared toward protocols that may one day become public as opposed to private ones (https://specs.openstack.org/openstack/api-wg/guidelines/headers.html).

In any case, going with a more specific prefix (like `VIP-`) is probably better. Also, this header will be going away :)

Reply:
X-Reason was removed.